core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -303,15 +311,25 @@ class LogSegment(val log: FileRecords,
     // We may need to reload the max timestamp after truncation.
     if (maxTimestampSoFar >= 0)
       loadLargestTimestamp(readToLogEnd = true)
+    _nextOffset = math.max(baseOffset, math.min(_nextOffset, offset))
     bytesTruncated
   }
 
   /**
-   * Calculate the offset that would be used for the next message to be append to this segment.
+   * The offset that would be used for the next message to be appended to this segment
+   */
+  def nextOffset(): Long = {
+    if (_nextOffset == INVALID_NEXT_OFFSET)
+      _nextOffset = nextOffsetFromLog()
```

Comment:
I get your concern. I am not sure we could just use the base offset of the next segment, because `nextOffset()` is a method that exists prior to this KIP and is actually used by `Log` to initialized `nextOffsetMetadata` (i.e. `nextOffsetMetadata = new LogOffsetMetadata(activeSegment.nextOffset(), activeSegment.baseOffset, activeSegment.size.toInt)`). Thus we actually need `nextOffset` of the active log segment.

I have updated the patch so that `nextOffset()` is a read-only operation and only returns the cached next offset. And I make sure that for non-empty segment that is read from disk, we will always call `updateNextOffsetFromLog()` before the first call to its `nextOffset()`. I think it should resolve the problem.